### DISCUSSION TITLE
Persist middleware: inital value overwritting stored value

### DISCUSSION THREAD

**Discussion by gugahoi** [pmndrs/zustand#1077] [Q&A]
When using the `persist` middleware, I can see that the default values I provide to the store are always immediately set on load even if there already is data persisted. How can I leverage it so that persisted values are not overwritten by initial values?

For reference I am trying to use Capacitor Storage as my Storage Engine.

```ts
import { Storage } from '@capacitor/storage'
import create from 'zustand'
import { persist } from 'zustand/middleware'

type QuickSticksStoreState = {
  signedUp: boolean 
  setSignedUp: (value: boolean) => void
}

export const useQuickSticksStore = create(
  persist<QuickSticksStoreState>(
    (set) => ({
      signedUp: false,
      setSignedUp: (value: boolean) => {
        set((state) => ({ ...state, signedUp: value }))
      },
    }),
    {
      name: 'quick-sticks',
      getStorage: () => ({
        getItem: async (key) => {
          console.warn('Getting Item from local storage: ' + key)
          const { value } = await Storage.get({ key })
          console.warn(`Got Item from local storage: ${key} - ${value}`)
          return value
        },
        setItem: async (key, value) => {
          console.warn('Updating item in local storage: ' + JSON.stringify({ key, value }))
          await Storage.set({ key, value: JSON.stringify(value) })
        },
        removeItem: (key) => {
          console.warn('Removing item from local storage: ' + key)
          Storage.remove({ key })
        },
      }),
      onRehydrateStorage: (state) => {
        console.warn('Hydration started', state)
        return (state, error) => {
          console.warn('Hydration Fn', state, error)
        }
      },
    },
  ),
)
```

On load, the following are the logs from an iOS device:
```
⚡️  [warn] - Hydration started undefined
⚡️  [warn] - Getting Item from local storage: quick-sticks
⚡️  [warn] - Got Item from local storage: quick-sticks - "{\"state\":{\"signedUp\":true},\"version\":0}"
⚡️  [warn] - Updating item in local storage: {"key":"quick-sticks","value":"{\"state\":{\"signedUp\":false},\"version\":0}"}
⚡️  [warn] - Hydration Fn {"signedUp":false} undefined
```

I don't understand why `setItem` is getting called immediately, before I ever invoke myself?

As you can see in the logs, it first read the persisted state correctly with `signedUp: true` but immediately overwrote it to `signedUp: false`.

Thanks in advance.

**Comment by AnatoleLucet** [comment]
Hi :wave: could you provide a codesandbox/repo to reproduce? :pray: 
I'm unsure how to setup `@capacitor/storage` (installing the package doesn't seem to be enough)

**Reply by gugahoi** [reply]
Thanks for checking in, I've created a minimum reproducible repo here: https://github.com/gugahoi/zustand-persist-capacitor however the problem does not happen. From testing I think the culprit was the fact that I was manually serializing the payload: `await Storage.set({ key, value: JSON.stringify(value) })` <- removing that `JSON.stringify` makes everything work.
